File: pipeline/reimpl/python/cotrn01c.py
"""
Reimplementation of COTRN01C — CardDemo Transaction View Screen.

CICS online program. Displays details of a single transaction by ID.
Transaction ID may be passed via commarea (from list screen selection)
or typed by the user.

Navigation:
  ENTER  → look up transaction by ID
  PF3    → return to calling program
  PF4    → clear screen
  PF5    → return to transaction list (COTRN00C)
"""

from __future__ import annotations
from dataclasses import dataclass
from decimal import Decimal
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, TranRecord,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3, DFHPF4, DFHPF5,
)


WS_PGMNAME = "COTRN01C"
WS_TRANID  = "CT01"


@dataclass
class TranViewResult:
    tran_record: Optional[TranRecord] = None
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    xctl_program: str = ""
    return_to_prev: bool = False
    cleared: bool = False


class TranRepository:
    def __init__(self, transactions: dict[str, TranRecord]):
        self._trans = transactions  # keyed by tran_id

    def find(self, tran_id: str) -> Optional[TranRecord]:
        return self._trans.get(tran_id.strip())


def process_tran_view(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    tran_id_input: str,
    tran_repo: TranRepository,
    preloaded_tran_id: str = "",
) -> TranViewResult:
    """Process transaction view screen — mirrors COTRN01C PROCEDURE DIVISION."""
    result = TranViewResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        if preloaded_tran_id and preloaded_tran_id.strip():
            _lookup_tran(preloaded_tran_id, tran_repo, result)
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    if eibaid in (DFHENTER, "ENTER"):
        return _process_enter(tran_id_input, tran_repo, commarea, result)
    elif eibaid in (DFHPF3, "PF3"):
        back = commarea.cdemo_from_program or "COMEN01C"
        result.xctl_program = back
        result.return_to_prev = True
        result.commarea = commarea
        return result
    elif eibaid in (DFHPF4, "PF4"):
        result.cleared = True
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.commarea = commarea
        return result
    elif eibaid in (DFHPF5, "PF5"):
        result.xctl_program = "COTRN00C"
        result.return_to_prev = True
        result.commarea = commarea
        return result
    else:
        result.error = True
        result.message = CCDA_MSG_INVALID_KEY
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result


def _process_enter(tran_id_input, repo, commarea, result):
    if not tran_id_input or tran_id_input.strip() == "":
        result.error = True
        result.message = "Tran ID can NOT be empty..."
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    _lookup_tran(tran_id_input, repo, result)
    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result


def _lookup_tran(tran_id, repo, result):
    tran = repo.find(tran_id)
    if tran is None:
        result.error = True
        result.message = "Transaction ID NOT found..."
    else:
        result.tran_record = tran
        result.message = ""


# ── Differential harness runner adapter ──────────────────────────────────────

_SEED_TRANS = {
    "0000000000000001": TranRecord(
        tran_id="0000000000000001", tran_type_cd="01", tran_cat_cd=1,
        tran_source="ONLINE", tran_desc="PURCHASE AT STORE",
        tran_amt=Decimal("125.50"), tran_card_num="4111111111111111",
        tran_merchant_id=100001, tran_merchant_name="ACME STORE",
        tran_merchant_city="NEW YORK", tran_merchant_zip="10001",
        tran_orig_ts="2025-01-10-09.00.00.000000",
        tran_proc_ts="2025-01-10-09.05.00.000000",
    ),
    "0000000000000002": TranRecord(
        tran_id="0000000000000002", tran_type_cd="02", tran_cat_cd=2,
        tran_source="POS TERM", tran_desc="BILL PAYMENT",
        tran_amt=Decimal("500.00"), tran_card_num="4222222222222222",
        tran_merchant_id=100002, tran_merchant_name="UTILITY CO",
        tran_merchant_city="CHICAGO", tran_merchant_zip="60601",
        tran_orig_ts="2025-01-12-14.30.00.000000",
        tran_proc_ts="2025-01-12-14.35.00.000000",
    ),
}


def run_vector(inputs: dict) -> dict:
    """Canonical runner entry point for the differential harness.

    SCENARIO selects a hardcoded test path:
      VIEW_FOUND      — look up existing transaction → display details
      VIEW_NOT_FOUND  — look up non-existent transaction → error
      EMPTY_TRAN_ID   — empty transaction ID → error
      INVALID_KEY     — press PF9 → invalid key error
      PF3_RETURN      — press PF3 → return to calling program
    """
    scenario = inputs.get("SCENARIO", "VIEW_FOUND")

    repo = TranRepository(dict(_SEED_TRANS))

    commarea = CarddemoCommarea(
        cdemo_from_tranid="CM00",
        cdemo_from_program="COMEN01C",
        cdemo_user_id="USER0001",
        cdemo_user_type="U",
        cdemo_pgm_context=1,
    )

    if scenario == "VIEW_FOUND":
        result = process_tran_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_id_input="0000000000000001", tran_repo=repo,
        )
    elif scenario == "VIEW_NOT_FOUND":
        result = process_tran_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_id_input="9999999999999999", tran_repo=repo,
        )
    elif scenario == "EMPTY_TRAN_ID":
        result = process_tran_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_id_input="", tran_repo=repo,
        )
    elif scenario == "INVALID_KEY":
        result = process_tran_view(
            eibcalen=100, eibaid="PF9", commarea=commarea,
            tran_id_input="", tran_repo=repo,
        )
    elif scenario == "PF3_RETURN":
        result = process_tran_view(
            eibcalen=100, eibaid="PF3", commarea=commarea,
            tran_id_input="", tran_repo=repo,
        )
    else:
        result = process_tran_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            tran_id_input="0000000000000001", tran_repo=repo,
        )

    tran_id = ""
    tran_type = ""
    tran_amt = ""
    tran_desc = ""
    tran_card = ""
    tran_merchant = ""
    if result.tran_record:
        tran_id = result.tran_record.tran_id
        tran_type = result.tran_record.tran_type_cd
        tran_amt = f"{result.tran_record.tran_amt:.2f}"
        tran_desc = result.tran_record.tran_desc
        tran_card = result.tran_record.tran_card_num
        tran_merchant = result.tran_record.tran_merchant_name

    return {
        "ERROR": "Y" if result.error else "N",
        "MESSAGE": result.message,
        "XCTL_PROGRAM": result.xctl_program or "",
        "TRAN_ID": tran_id,
        "TRAN_TYPE": tran_type,
        "TRAN_AMT": tran_amt,
        "TRAN_DESC": tran_desc,
        "TRAN_CARD": tran_card,
        "TRAN_MERCHANT": tran_merchant,
    }

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cotrn01c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Java reimplementation of COTRN01C — CardDemo Transaction View Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/cotrn01c.py}. Python is the source of truth.
 *
 * <p>Standalone business method: {@link #processTranView}. The {@link #runVector}
 * method is a thin adapter that builds scenario-specific parameters and delegates.
 */
public class Cotrn01c implements ProgramRunner {

    private static final String CCDA_MSG_INVALID_KEY = "Invalid key pressed.                    ";

    // ── Inner data types ─────────────────────────────────────────────────

    static class TranRecord {
        final String tranId;
        final String tranTypeCd;
        final String tranAmt;
        final String tranDesc;
        final String tranCardNum;
        final String tranMerchantName;

        TranRecord(String tranId, String tranTypeCd, String tranAmt,
                   String tranDesc, String tranCardNum, String tranMerchantName) {
            this.tranId = tranId;
            this.tranTypeCd = tranTypeCd;
            this.tranAmt = tranAmt;
            this.tranDesc = tranDesc;
            this.tranCardNum = tranCardNum;
            this.tranMerchantName = tranMerchantName;
        }
    }

    static class TranViewResult {
        TranRecord tranRecord = null;
        boolean error = false;
        String message = "";
        String xctlProgram = "";
        boolean cleared = false;
    }

    static class TranRepo {
        private final Map<String, TranRecord> trans;
        TranRepo(Map<String, TranRecord> trans) { this.trans = trans; }

        TranRecord find(String tranId) {
            return trans.get(tranId.trim());
        }
    }

    // ── Standalone business method ──────────────────────────────────────

    /**
     * Process transaction view screen — mirrors Python's process_tran_view.
     *
     * @param eibcalen        CICS EIBCALEN
     * @param eibaid          aid key pressed
     * @param pgmContext      commarea program context (0 = initial)
     * @param fromProgram     commarea cdemo_from_program
     * @param tranIdInput     transaction ID typed by user
     * @param preloadedTranId transaction ID preloaded from previous screen
     * @param tranRepo        transaction repository
     * @return TranViewResult with all output fields
     */
    TranViewResult processTranView(
            int eibcalen,
            String eibaid,
            int pgmContext,
            String fromProgram,
            String tranIdInput,
            String preloadedTranId,
            TranRepo tranRepo
    ) {
        TranViewResult result = new TranViewResult();

        if (eibcalen == 0) {
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (pgmContext == 0) {
            if (preloadedTranId != null && !preloadedTranId.trim().isEmpty()) {
                lookupTran(preloadedTranId, tranRepo, result);
            }
            return result;
        }

        switch (eibaid) {
            case "ENTER":
                return processEnter(tranIdInput, tranRepo, result);
            case "PF3": {
                String back = (fromProgram != null && !fromProgram.isEmpty()) ? fromProgram : "COMEN01C";
                result.xctlProgram = back;
                return result;
            }
            case "PF4":
                result.cleared = true;
                return result;
            case "PF5":
                result.xctlProgram = "COTRN00C";
                return result;
            default:
                result.error = true;
                result.message = CCDA_MSG_INVALID_KEY;
                return result;
        }
    }

    private TranViewResult processEnter(String tranIdInput, TranRepo repo, TranViewResult result) {
        if (tranIdInput == null || tranIdInput.trim().isEmpty()) {
            result.error = true;
            result.message = "Tran ID can NOT be empty...";
            return result;
        }
        lookupTran(tranIdInput, repo, result);
        return result;
    }

    private void lookupTran(String tranId, TranRepo repo, TranViewResult result) {
        TranRecord tran = repo.find(tranId);
        if (tran == null) {
            result.error = true;
            result.message = "Transaction ID NOT found...";
        } else {
            result.tranRecord = tran;
            result.message = "";
        }
    }

    // ── Seed data ───────────────────────────────────────────────────────

    private static final Map<String, TranRecord> SEED_TRANS = new LinkedHashMap<>();
    static {
        SEED_TRANS.put("0000000000000001", new TranRecord(
                "0000000000000001", "01", "125.50", "PURCHASE AT STORE",
                "4111111111111111", "ACME STORE"));
        SEED_TRANS.put("0000000000000002", new TranRecord(
                "0000000000000002", "02", "500.00", "BILL PAYMENT",
                "4222222222222222", "UTILITY CO"));
    }

    // ── Thin runVector adapter ──────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "VIEW_FOUND");

        TranRepo repo = new TranRepo(new LinkedHashMap<>(SEED_TRANS));

        TranViewResult result;

        switch (scenario) {
            case "VIEW_FOUND":
                result = processTranView(100, "ENTER", 1, "COMEN01C",
                        "0000000000000001", "", repo);
                break;

            case "VIEW_NOT_FOUND":
                result = processTranView(100, "ENTER", 1, "COMEN01C",
                        "9999999999999999", "", repo);
                break;

            case "EMPTY_TRAN_ID":
                result = processTranView(100, "ENTER", 1, "COMEN01C",
                        "", "", repo);
                break;

            case "INVALID_KEY":
                result = processTranView(100, "PF9", 1, "COMEN01C",
                        "", "", repo);
                break;

            case "PF3_RETURN":
                result = processTranView(100, "PF3", 1, "COMEN01C",
                        "", "", repo);
                break;

            default:
                result = processTranView(100, "ENTER", 1, "COMEN01C",
                        "0000000000000001", "", repo);
                break;
        }

        String tranId = "";
        String tranType = "";
        String tranAmt = "";
        String tranDesc = "";
        String tranCard = "";
        String tranMerchant = "";
        if (result.tranRecord != null) {
            tranId = result.tranRecord.tranId;
            tranType = result.tranRecord.tranTypeCd;
            tranAmt = result.tranRecord.tranAmt;
            tranDesc = result.tranRecord.tranDesc;
            tranCard = result.tranRecord.tranCardNum;
            tranMerchant = result.tranRecord.tranMerchantName;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram);
        out.put("TRAN_ID", tranId);
        out.put("TRAN_TYPE", tranType);
        out.put("TRAN_AMT", tranAmt);
        out.put("TRAN_DESC", tranDesc);
        out.put("TRAN_CARD", tranCard);
        out.put("TRAN_MERCHANT", tranMerchant);
        return out;
    }
}
